src/Compilers/CSharp/Portable/Lowering/LocalRewriter/LocalRewriter_StringConcat.cs

```diff
@@ -336,16 +476,108 @@ private BoundExpression RewriteStringConcatenationFourExprs(SyntaxNode syntax, B
         private BoundExpression RewriteStringConcatenationManyExprs(SyntaxNode syntax, ImmutableArray<BoundExpression> loweredArgs)
         {
             Debug.Assert(loweredArgs.Length > 4);
-            Debug.Assert(loweredArgs.All(a => a.HasErrors || a.Type is { } && a.Type.IsStringType()));
+
+            var stringifiedArgs = ArrayBuilder<BoundExpression>.GetInstance(loweredArgs.Length);
+
+            foreach (var loweredArg in loweredArgs)
+            {
+                Debug.Assert(loweredArg.HasErrors || loweredArg.Type is { } argType && (argType.IsStringType() || argType.IsCharType()));
```

Comment:
No, it would not. We just won't unwrap concat at the last iteration and get `string.Concat(string.Concat(new ReadOnlySpan<char>(in c), s, s, s), s)` at the end. Test `ConcatFive_Char` already verifies that